### ISSUE TITLE
Use of <includeStdTypes> breaks <protocCommand>

### ISSUE THREAD

**Issue by alanbur** [os72/protoc-jar-maven-plugin#40]
I'm using protoc-jar-maven-plugin to build cross-platform and need to use it on a platform that it doesn't support directly, although protoc itself is available for the platform. I'm using the `<protocCommand>` tag in the pom.xml to reference the external protoc executable, which all works fine.

The issue is if I also need to use `<includeStdTypes>true</includeStdTypes>` at the same time. Doing so generates the following error:

`[ERROR] Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.2.0:run (default) on project higgs-management-protocol: Error extracting protoc for version 320: Unsupported platform: sunos/amd64 -> [Help 1]`

So why is the plugin trying to extract a copy of protoc when it has been told to use an external version of protoc?

The issue seems to be this line of code:

[ProtocJarMojo.java:311](https://github.com/os72/protoc-jar-maven-plugin/blob/master/src/main/java/com/github/os72/protocjar/maven/ProtocJarMojo.java#L311)

`if ((protocCommand == null && protocArtifact == null) || includeStdTypes) {`

In this case `includeStdTypes` is true and the code block following attempts to extract protoc from the plugin JAR, and because the plugin doesn't directly support this platform, it fails to do so and exits with an error.

I believe the logic here is erroneous - if `protocCommand` and `includeStdTypes` are both set, no attempt to extract a protoc binary from the JAR should be made and the determination of the `-I` flag to be added at line 338 should be based on the absolute path of the external protoc command and not the location of the one that is normally extracted from the JAR.

**Comment by os72** [comment]
Ack. Would you be able to build SNAPSHOT and verify a fix?

**Comment by alanbur** [comment]
I'd be happy to build and verify a fix but I'll need some help doing so, are there instructions on how to do that? I assume it's not entirely straightforward because of the need to embed the linux, win & osx protoc binaries inside the JAR.

**Comment by os72** [comment]
No, protoc is provided, just clone protoc-jar and protoc-jar-maven-plugin, and maven clean install, that's it

**Comment by alanbur** [comment]
It won't build at all on Solaris, and on Linux I get the following errors building protoc-jar-maven-plugin:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.9.1:jar (attach-javadoc) on project protoc-jar-maven-plugin: MavenReportException: Error while creating archive: 
[ERROR] Exit code: 1 - /tmp/protoc-jar-maven-plugin/src/main/java/com/github/os72/protocjar/maven/ProtocJarMojo.java:52: error: unknown tag: goal
[ERROR]  * @goal run
[ERROR]    ^
[ERROR] /tmp/protoc-jar-maven-plugin/src/main/java/com/github/os72/protocjar/maven/ProtocJarMojo.java:53: error: unknown tag: phase
[ERROR]  * @phase generate-sources
[ERROR]    ^
[ERROR] /tmp/protoc-jar-maven-plugin/src/main/java/com/github/os72/protocjar/maven/ProtocJarMojo.java:54: error: unknown tag: requiresDependencyResolution
[ERROR]  * @requiresDependencyResolution
[ERROR]    ^
[ERROR] 
[ERROR] Command line was: /usr/lib/jvm/jdk-8-oracle-x64/jre/../bin/javadoc @options @packages
[ERROR] 
[ERROR] Refer to the generated Javadoc files in '/tmp/protoc-jar-maven-plugin/target/apidocs' dir.
```

If I comment the offending lines out I still get an error:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-plugin-plugin:3.2:descriptor (default-descriptor) on project protoc-jar-maven-plugin: Error extracting plugin descriptor: 'No mojo definitions were found for plugin: com.github.os72:protoc-jar-maven-plugin.' -> [Help 1]
```

**Comment by alanbur** [comment]
As the issue above only seems to affect the building of the documentation JAR I manually installed the protoc-jar-maven-plugin JAR into the Maven cache and tested with that. It doesn't fail any more but I'm not convinced it is 100% correct:

```
protoc-jar: executing: [protoc-3.2.0, -I/var/tmp/protocjar8125148272656799685/include, -I/pool/home/alanbur/higgs/higgs-controlplane/higgs-management-protocol/src/main/protobuf, -I/pool/home/alanbur/higgs/higgs-controlplane/higgs-management-protocol/src/main/protobuf, --java_out=/pool/home/alanbur/higgs/higgs-controlplane/higgs-management-protocol/target/generated-sources, /pool/home/alanbur/higgs/higgs-controlplane/higgs-management-protocol/src/main/protobuf/manage_agent_v1.proto]
```

I believe the first `-I` should reference the include directory that's relative to the executed protoc binary, not to a copy of the protoc include files that have been extracted from the JAR file, otherwise there is a potential for version mismatch between the protoc executable and the include files.

**Comment by os72** [comment]
Ah if you have includes along with your binary that requires no fix and no `includeStdTypes`. Just additional `includeDirectories`
```
<includeDirectories>
	<include>[/usr/lib/protobuf or wherever]/include</include>
</includeDirectories>
```

Will this work?

I see now what you meant - with `protocCommand` and `includeStdTypes` we could try to infer the `include` location from the `protoc` path. But I'm not sure how robust that would be. What's the path structure for you?

**Comment by alanbur** [comment]
The problem with `<includeDirectories>` is that it requires a specific path but the location of the protoc standard includes varies from platform to platform, `<includeStdTypes>` is actually the correct solution - **if** it can be made to work reliably, as you say...

In my case the path structure for protoc has `bin`, `include` and `lib` directories all at the same level, so relative to the binary the include files are at `../include`. That's because I'm using a custom build of protoc. If I look on a random Linux machine (Debian-based) that relationship doesn't seem to hold, in fact it doesn't seem to ship the standard protoc header files at all, unfortunately :-(

I think you are most probably right and there's no reliable way of doing this in the maven plugin. There should probably be a `-includeStdTypes` in protobuf itself, but of course fixing **that** is well outside of the scope of this fix. As such, I think your current fix is the best solution that can be achieved in the circumstances.

Thanks for responding to this so quickly, it's much appreciated :-)

**Comment by alanbur** [comment]
p.s. This fix is obviously against the tip of the 3.x protobuf series, will the fix be propagated back into the 2.6.x chain as well? Thanks!

**Comment by os72** [comment]
Yes `includeStdTypes` populates includes based on given `protocVersion`, 2.x or 3.x. Note your build is version-specific anyway, you need to specify a matching `protobuf-java` version

In fact there is an old enhancement request #19 that could solve this problem differently. Haven't looked to implement it yet..

**Comment by os72** [comment]
Released 3.3.0.1

**Comment by alanbur** [comment]
Thanks! :+1:
